### PR TITLE
Do not copy images while rendering Adventure Map

### PR DESCRIPTION
**Pull request by ihhub**
relates to #5716

### CODE REVIEW

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 1]
```diff
@@ -51,112 +51,82 @@
...
-
-    void populateStaticTileUnfitObjectInfo( TileUnfitRenderObjectInfo & tileUnfit, std::vector<std::pair<fheroes2::Point, fheroes2::Sprite>> & imageInfo,
-                                            std::vector<std::pair<fheroes2::Point, fheroes2::Sprite>> & shadowInfo, const fheroes2::Point & offset,
-                                            const uint8_t alphaValue )
-    {
-        for ( auto & [imagePos, image] : imageInfo ) {
             if ( imagePos.y > 0 ) {
                 if ( imagePos.x < 0 ) {
-                    tileUnfit.bottomBackgroundImages[imagePos + offset].emplace_front( std::move( image ), alphaValue );
+                    tileUnfit.bottomBackgroundImages[imagePos + offset].emplace_front( std::move( objectInfo ) );
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/gui/interface_gamearea.cpp:82]: Hi @ihhub we can add the [compile-type check](https://en.cppreference.com/w/cpp/types/is_trivially_copyable) whether the `objectInfo` is trivially copyable and remove these redundant `std::move()` calls.

File: src/engine/image.cpp [hunk 2]
```diff
@@ -790,7 +790,7 @@ namespace fheroes2
         const uint8_t * gamePalette = getGamePalette();
 
         if ( flip ) {
-            const int32_t offsetInY = inY * widthIn + widthIn - 1 - inX;
+            const int32_t offsetInY = inY * widthIn + width - 1 + inX;
```
Comment (Districh-ru) [inline src/engine/image.cpp:793]: It is better to revert this logic as these function is widely used in the code and we will need to correct many places in the code then.
Also the original logic here was made to make the same calculation for flipped and non-flipped sprites render regions.
```suggestion
            const int32_t offsetInY = inY * widthIn + widthIn - 1 - inX;
```

File: src/engine/image.cpp [hunk 3]
```diff
@@ -993,7 +993,7 @@ namespace fheroes2
         const int32_t widthOut = out.width();
 
         if ( flip ) {
-            const int32_t offsetInY = inY * widthIn + widthIn - 1 - inX;
+            const int32_t offsetInY = inY * widthIn + width - 1 + inX;
```
Comment (Districh-ru) [inline src/engine/image.cpp:996]: The same as for 793 line:
```suggestion
            const int32_t offsetInY = inY * widthIn + widthIn - 1 - inX;
```

**Review by Districh-ru — CHANGES_REQUESTED** [review]
Hi @ihhub! I left a couple suggestions to fix rendering of flipped images. Could you take a look when you have time.

File: src/engine/image.cpp [hunk 4]
```diff
@@ -1562,22 +1562,17 @@ namespace fheroes2
...
-                        = Crop( original, intersection.x - spriteRelativeOffset.x, intersection.y - spriteRelativeOffset.y, intersection.width, intersection.height );
-                    assert( !cropped.empty() );
-                    cropped.setPosition( intersection.x - roi.x, intersection.y - roi.y );
+                    outputSquareId.emplace_back( offset + Point( x, y ) );
 
-                    output.emplace_back( offset + Point( x, y ), std::move( cropped ) );
+                    outputImageInfo.emplace_back( fheroes2::Point( intersection.x - roi.x, intersection.y - roi.y ),
+                                                  fheroes2::Rect( intersection.x - spriteRelativeOffset.x, intersection.y - spriteRelativeOffset.y, intersection.width,
+                                                                  intersection.height ) );
                 }
```
Comment (Districh-ru) [inline src/engine/image.cpp:1576]: As now we are using `Blit` later and here are doing only coordinates calculation we do not need to use `const bool flip` and make the same calculations for flipped and non-flipped sprites:
```cpp
        for ( int32_t y = 0; y < stepPerDirection.y; ++y ) {
            for ( int32_t x = 0; x < stepPerDirection.x; ++x ) {
                const Rect roi( x * squareSize, y * squareSize, squareSize, squareSize );
                const Rect intersection = relativeROI ^ roi;
                assert( intersection.width > 0 && intersection.height > 0 );

                outputSquareId.emplace_back( offset + Point( x, y ) );

                outputImageInfo.emplace_back( fheroes2::Point( intersection.x - roi.x, intersection.y - roi.y ),
                                              fheroes2::Rect( intersection.x - spriteRelativeOffset.x, intersection.y - spriteRelativeOffset.y, intersection.width,
                                                              intersection.height ) );
            }
        }
```

**Comment by oleg-derevenetz** [comment]
Hi @ihhub please address IWYU suggestions :)

**Comment by ihhub** [comment]
> Hi @ihhub please address IWYU suggestions :)

As usual :)

**Review by oleg-derevenetz — COMMENTED** [review]
Hi @ihhub I just left one question, could you please take a look when possible?

File: src/fheroes2/gui/interface_gamearea.cpp [hunk 5]
```diff
@@ -51,112 +52,84 @@
...
-            , alphaValue( value )
-        {
-            // Do nothing.
-        }
-
-        fheroes2::Sprite image;
-
-        uint8_t alphaValue{ 255 };
-    };
+    static_assert( std::is_trivially_copyable<fheroes2::ObjectRenderingInfo>::value, "This class is not trivially copyable anymore. Add std::move where required." );
```
[the remark this answers is not in the document]
Comment (oleg-derevenetz) [reply src/fheroes2/gui/interface_gamearea.cpp:62]: Ah, I got the idea.

**Review by oleg-derevenetz — APPROVED** [review]

File: src/fheroes2/gui/interface_gamearea.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/gui/interface_gamearea.cpp:62]: It was initially proposed by you :)

**Review by Districh-ru — APPROVED** [review]
I played fully one M map (also using boats) and did not notice rendering bugs.

**Comment by Branikolog** [comment]
I also haven't found any new issues.

**Pull request merged**